A27-xds-global-load-balancing.md

```diff
@@ -0,0 +1,442 @@
+Note that the EDS policy will not support per-endpoint stats; it will
+report only per-locality stats.  Currently, the EDS policy will report
+only client-side stats; in the future, we will support also reporting
+backend server stats reported to the client via
+[ORCA](https://docs.google.com/document/d/1NSnK3346BkBo1JUU3I9I5NYYnaJZQPt8_Z_XCBCI3uA/edit).
+
+### How gRPC Will Use the xDS API
+
+gRPC does not have exactly the same set of features as Envoy, so there
+are by necessity some small differences between how the two clients use
+the xDS API.  This section documents exactly what gRPC requires from the
+xDS server and what xDS features it does and does not support.
+
+Because there are many fields in the xDS APIs that gRPC will not use in
+its initial version, the gRPC client must in general ignore any fields
+it does not support.  This does mean that if we add support for one of
+these fields in a future version and users enable use of the feature in
+the management server, older clients will continue to ignore the field,
+so they may not actually get the new behavior.  However, the alternative
+is even worse: if we make the client strict (i.e., it will fail if the
+management server populates fields that the client does not yet support),
+then the same scenario will cause the older clients to fail instead of
+just ignoring the new feature.
+
+Note, however, that there are some unavoidable cases where enabling a
+new feature will cause older clients to break.  These specific cases
+will be called out below.
+
+#### Initial Request on the ADS Stream
+
+The first request on the ADS stream will include a Node message that
+identifies the client.  As mentioned above, most of the fields for the Node
+message will be read from the bootstrap file.
+
+#### LDS
+
+The gRPC client will typically start by sending an LDS request for a
+`Listener` resource whose name matches the server name from the target
+URI used to create the gRPC channel (including port suffix if present).
+Note that the server name will also be used in the HTTP/2 ":authority"
+field, so it needs to be in host or host:port syntax.  However, the gRPC
+client will not interpret it in any special way.  This means that the
+xDS server effectively defines what the default port means if the URI
+does not specify a port.
+
+Because we are requesting one specific resource by name, the LDS response
+should include at least one `Listener`.  However, many existing xDS servers
+do not support requesting one specific resource in LDS, so the gRPC client
+must be tolerant of servers that may ignore the requested resource name;
+if the server returns multiple resources, the client will look for the one
+with the name it asked for, and it will ignore the rest of the entries.
+Note that this means that the resource returned by the xDS server must
+have exactly the name specified by the client.
```

Comment:
Given the above point you make about requiring that management servers be API listener aware, why not also ask them to support LDS selection hints?

Reply:
FWIW, I don't think this buys us anything, but it's a bit inconsistent on what can/can't be asked of the management server to the reader.

Reply:
The intention here was to try to minimize the set of changes that existing management servers need to make to support gRPC.  Existing management servers only support wildcard queries for LDS requests, because that's what Envoy does.  So while it's true that management servers will need to add support for API listeners, they don't need to support queries for specific resource names if they don't want to.

It may very well be that this is of limited value, since it's not really hard to support queries for specific resource names.  But it also wasn't hard to support this in the client either.

In general, I do certainly agree that the semantics of selecting which resources the client is interested in are fairly messy in xDS.  I think this is something we need to continue to discuss going forward.